clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java

```diff
@@ -68,6 +70,15 @@ public int partition(String topic, Object key, byte[] keyBytes, Object value, by
         }
     }
 
+    public int getNextValue(String topic) {
+        AtomicInteger counter = topicCounterMap.get(topic);
+        if (null == counter) {
```

Comment:
yeah, I was aware of it and chose to do it in this way. 
As this method is a hot method which may be called by every producer.send() if no key is provided, put it as 'synchronized' may potentially impact the performance among the threads. 

while the null check may be both true for the initial concurrent threads, as you said, it only impact that very first few calls. further calls are not impacted.